src/BillingAddressForm.tsx

```diff
@@ -0,0 +1,248 @@
+  }, [firstName, lastName, address1, city, zip, country, state]);
+
+  const hasErrors =
+    firstNameError ||
+    lastNameError ||
+    address1Error ||
+    cityError ||
+    zipError ||
+    countryError ||
+    stateError;
+
+  const hasEmptyRequiredFields = () => {
+    const isEmpty = (value: string) => value === "";
+    return (
+      isEmpty(firstName) ||
+      isEmpty(lastName) ||
+      isEmpty(address1) ||
+      isEmpty(city) ||
+      isEmpty(zip) ||
+      isEmpty(country) ||
+      isEmpty(state)
+    );
+  };
+
+  const checkFormErrors = () => {
+    if (hasEmptyRequiredFields() || hasErrors) {
+      onFormError();
+    } else {
+      onFormErrorResolved();
+    }
+  };
+
+  const handleFirstNameChange = (
+    event: React.ChangeEvent<HTMLInputElement>
+  ) => {
+    const value = event.target.value;
+    const isEmpty = value === "";
+    setFirstNameError(isEmpty);
+    setFirstName(value);
+  };
+
+  const handleLastNameChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    const value = event.target.value;
+    const isEmpty = value === "";
+    setLastNameError(isEmpty);
+    setLastName(value);
+  };
+
+  const handleAddress1Change = (event: React.ChangeEvent<HTMLInputElement>) => {
+    const value = event.target.value;
+    const isEmpty = value === "";
+    setAddress1Error(isEmpty);
+    setAddress1(value);
+  };
+
+  const handleCityChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    const value = event.target.value;
+    const isEmpty = value === "";
+    setCityError(isEmpty);
+    setCity(value);
+  };
+
+  const handleAddress2Change = (event: React.ChangeEvent<HTMLInputElement>) => {
+    const value = event.target.value;
+    setAddress2(value);
+  };
+
+  const handleStateChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    const value = event.target.value;
+    const isEmpty = value === "";
+    setStateError(isEmpty);
+    setState(value);
+  };
+
+  const handleZipChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    const value = event.target.value;
+    const isNotFiveDigits = !/^\d{5}$/.test(value);
+    setZipError(isNotFiveDigits || value === "");
+    setZip(value);
+  };
+
+  const handleCountryChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    const value = event.target.value;
+    const isEmpty = value === "";
+    setCountryError(isEmpty);
+    setCountry(value);
+  };
+
+  return (
+    <Box sx={{ flexGrow: 1 }}>
+      <Typography variant="h5" gutterBottom>
+        Billing Address
+      </Typography>
+      <Grid container spacing={3}>
+        <Grid xs={12} sm={6}>
+          <TextField
+            required
+            id="firstName"
+            name="firstName"
+            label="First name"
+            fullWidth
+            autoComplete="given-name"
+            variant="standard"
+            error={firstNameError}
+            onChange={handleFirstNameChange}
+            value={firstName}
+          />
```

Comment:
This is repeated a lot, so extracting it to a shared variable might be helpful. This applies to both Form components.
```suggestion
          <TextField
           ...commonProps
          />
```